solidity/contracts/isms/hook/OPL2ToL1Ism.sol

```diff
@@ -0,0 +1,111 @@
+import {AbstractMessageIdAuthorizedIsm} from "./AbstractMessageIdAuthorizedIsm.sol";
+
+// ============ External Imports ============
+
+import {ICrossDomainMessenger} from "../../interfaces/optimism/ICrossDomainMessenger.sol";
+import {IOptimismPortal} from "../../interfaces/optimism/IOptimismPortal.sol";
+import {CrossChainEnabledOptimism} from "@openzeppelin/contracts/crosschain/optimism/CrossChainEnabledOptimism.sol";
+import {Address} from "@openzeppelin/contracts/utils/Address.sol";
+
+/**
+ * @title OPL2ToL1Ism
+ * @notice Uses the native Optimism bridge to verify interchain messages from L2 to L1.
+ */
+contract OPL2ToL1Ism is
+    CrossChainEnabledOptimism,
+    AbstractMessageIdAuthorizedIsm
+{
+    using TypeCasts for address;
+    using Message for bytes;
+    using OPL2ToL1Metadata for bytes;
+
+    // ============ Constants ============
+
+    // module type for the ISM
+    uint8 public constant moduleType =
+        uint8(IInterchainSecurityModule.Types.OP_L2_TO_L1);
+    // OptimismPortal contract on L1 to finalize withdrawal from L1
+    IOptimismPortal public immutable portal;
+
+    // ============ Constructor ============
+
+    constructor(address _messenger) CrossChainEnabledOptimism(_messenger) {
+        address _portal = ICrossDomainMessenger(_messenger).PORTAL();
+        require(
+            Address.isContract(_portal),
+            "OPL2ToL1Ism: invalid OptimismPortal contract"
+        );
+        portal = IOptimismPortal(_portal);
+    }
+
+    // ============ External Functions ============
+
+    /// @inheritdoc IInterchainSecurityModule
+    function verify(
+        bytes calldata metadata,
+        bytes calldata message
+    ) external override returns (bool) {
+        bool verified = isVerified(message);
+        if (!verified) {
+            require(_verifyWithPortalCall(metadata, message));
+        }
+        releaseValueToRecipient(message);
+        return true;
+    }
+
+    // ============ Internal function ============
+
+    /**
+     * @notice Verify message directly using the portal.finalizeWithdrawal function.
+     * @dev This is a fallback in case the message is not verified by the stateful verify function first.
+     */
+    function _verifyWithPortalCall(
+        bytes calldata metadata,
+        bytes calldata message
+    ) internal returns (bool) {
+        require(
+            metadata.checkCalldataLength(),
+            "OPL2ToL1Ism: invalid data length"
+        );
+        require(
+            metadata.messageId() == message.id(),
+            "OPL2ToL1Ism: invalid message id"
+        );
+
+        IOptimismPortal.WithdrawalTransaction memory withdrawal = abi.decode(
+            metadata,
+            (IOptimismPortal.WithdrawalTransaction)
+        );
+        portal.finalizeWithdrawalTransaction(withdrawal);
+
+        // if the finalizeWithdrawalTransaction call is successful, the message is verified
+        return true;
```

Comment:
maybe you can remove the bool return value from this function and skip the `require` branch in the other `verify` function